src/target/riscv/batch.c

```diff
@@ -210,12 +232,22 @@ size_t riscv_batch_available_scans(struct riscv_batch *batch)
 
 bool riscv_batch_dmi_busy_encountered(const struct riscv_batch *batch)
 {
+	assert(batch->was_run);
 	if (batch->used_scans == 0)
 		/* Empty batch */
 		return false;
 
 	assert(batch->last_scan == RISCV_SCAN_TYPE_NOP);
-	const struct scan_field *field = batch->fields + batch->used_scans - 1;
-	const uint64_t in = buf_get_u64(field->in_value, 0, field->num_bits);
-	return get_field(in, DTM_DMI_OP) == DTM_DMI_OP_BUSY;
+	return riscv_batch_was_busy(batch, batch->used_scans - 1);
+}
+
+size_t riscv_batch_finished_scans(const struct riscv_batch *batch)
+{
+	if (!riscv_batch_dmi_busy_encountered(batch))
+		return batch->used_scans;
+	assert(batch->used_scans);
+	size_t first_busy = 0;
+	while (!riscv_batch_was_busy(batch, first_busy))
+		++first_busy;
+	return first_busy;
```

Comment:
Few cosmetic details - if you like:

```suggestion
	assert(batch->used_scans > 0);
	if (!riscv_batch_dmi_busy_encountered(batch))
		/* Whole batch succeeded */
		return batch->used_scans;

	size_t first_busy = 0;
	while (!riscv_batch_was_busy(batch, first_busy))
		++first_busy;
	return first_busy;
```

Reply:
Addressed.

Though I left `assert(batch->used_scans > 0)` where it was. It's related to the loop condition, and I would like to leave the assertion as close to what it guards as possible.